thread-rcu/rcu.h

```diff
@@ -0,0 +1,317 @@
+    })
+
+#define atomic_xchg_release(x, v)                                              \
+    ({                                                                         \
+        __typeof__(*x) __a_c_r_x;                                              \
+        atomic_exchange_explicit((volatile _Atomic __typeof__(__a_c_r_x) *) x, \
+                                 v, memory_order_release);                     \
+    })
+
+#include <errno.h>
+#include <stdatomic.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#ifdef __CHECKER__
+#define __rcu __attribute__((noderef, address_space(__rcu)))
+#define rcu_check_sparse(p, space) ((void) (((typeof(*p) space *) p) == p))
+#define __force __attribute__((force))
+#define rcu_uncheck(p) ((__typeof__(*(p)) __force *) (p))
+#define rcu_check(p) ((__typeof__(*(p)) __force __rcu *) (p))
+#else
+#define __rcu
+#define rcu_check_sparse(p, space)
+#define __force
+#define rcu_uncheck(p) p
+#define rcu_check(p) p
+#endif /* __CHECKER__ */
+
+/* Avoid false sharing */
+#define CACHE_LINE_SIZE 64
+#define __rcu_aligned __attribute__((aligned(2 * CACHE_LINE_SIZE)))
+
+/* Per-thread variable
+ *
+ * rcu_nesting is to distinguish where the current thread is in which grace
+ * period. The reader will use the global variable rcu_nesting_idx to set up
+ * the corresponding index of rcu_nesting for the read-side critical section.
+ * It is a per-thread variable for optimizing the reader-side lock execution.
+ *
+ * rcu_nesting uses two lowest bits of __next_rcu_nesting to determine the
+ * graced period.
+ */
+struct rcu_node {
+    unsigned int tid;
+    uintptr_t __next_rcu_nesting;
+} __rcu_aligned;
+
+struct rcu_data {
+    unsigned int nr_thread;
+    struct rcu_node *head;
+    unsigned int rcu_nesting_idx;
+    spinlock_t sp;
+};
+
+/* Easy to use */
+
+#define rcu_nesting(np, idx) \
+    (READ_ONCE((np)->__next_rcu_nesting) & (0x1 << ((idx) &0x1)))
+#define rcu_set_nesting(np, idx)                                          \
+    do {                                                                  \
+        WRITE_ONCE(                                                       \
+            (np)->__next_rcu_nesting,                                     \
+            READ_ONCE((np)->__next_rcu_nesting) | (0x1 << ((idx) &0x1))); \
+    } while (0)
+#define rcu_unset_nesting(np)                                          \
+    do {                                                               \
+        smp_store_release((np)->__next_rcu_nesting,                    \
+                          READ_ONCE((np)->__next_rcu_nesting) & ~0x3); \
+    } while (0)
+#define rcu_next(np) \
+    ((struct rcu_node *) (READ_ONCE((np)->__next_rcu_nesting) & ~0x3))
+#define rcu_next_mask(nrn) ((struct rcu_node *) ((uintptr_t) (nrn) & ~0x3))
+
+static struct rcu_data rcu_data = {
+    .nr_thread = 0,
+    .head = NULL,
+    .rcu_nesting_idx = 0,
```

Comment:
`rcu_nesting_idx` is a bit longer. Let's rename it to `nesting_idx`.

Reply:
`rcu_nesting_idx` is consist with `rcu_node->rcu_nesting`
(which actually is in the lowest two bits of `__next_rcu_nesting`).
I think the present name is satisfactory.